src/mcfret.c

```diff
@@ -0,0 +1,621 @@
+    FILE *mu_traj;
+    /* FILE *absorption_matrix, *emission_matrix,*/
+    FILE *log;
+    FILE *Cfile;
+
+    mu_xyz=(float *)calloc(non->singles*3,sizeof(float));
+    Hamil_i_e=(float *)calloc((non->singles+1)*non->singles/2,sizeof(float));
+
+    /* Open Trajectory files */
+    open_files(non,&H_traj,&mu_traj,&Cfile);
+
+    /* Here we want to call the routine for checking the trajectory files */ 
+    control(non);
+
+    /* Initialize sample numbers */
+    N_samples=determine_samples(non);
+    Ncl=0;
+
+    /* Read coupling, this is done if the coupling and transition-dipoles are */
+    /* time-independent and only one snapshot is stored */
+    read_coupling(non,C_traj,mu_traj,Hamil_i_e,mu_xyz);
+    samples=1;
+
+
+    log=fopen("NISE.log","a");
+    fprintf(log,"Begin sample: %d, End sample: %d.\n",non->begin,non->end);
+    fclose(log);
+
+
+    /* Looping over samples: Each sample represents a different starting point on the Hamiltonian trajectory */
+    for (samples=non->begin;samples<non->end;samples++){ 
+        ti=samples*non->sample;
+        if (non->cluster!=-1){
+            if (read_cluster(non,ti,&cl,Cfile)!=1){
+	              printf("Cluster trajectory file to short, could not fill buffer!!!\n");
+	              printf("ITIME %d\n",ti);
+	              exit(1);
+            }
+            /* Configuration belong to cluster */ 
+            if (non->cluster==cl){
+	              Ncl++;
+            }
+        }
+        if (non->cluster==-1 || non->cluster==cl){
+            /* Read Hamiltonian */
+            read_Hamiltonian(non,Hamil_i_e,H_traj,ti);
+          	    
+            /* Extract couplings between segments */
+            multi_projection_Coupling(Hamil_i_e,non);
+            for (i=0;i<non->singles;i++){
+              for (j=i+1;j<non->singles;j++){
+                J[non->singles*i+j]+=Hamil_i_e[Sindex(i,j,non->singles)];
+                J[non->singles*j+i]+=Hamil_i_e[Sindex(i,j,non->singles)];
+              }
+            }
+        } /* We are closing the cluster loop */
+
+         /* Update NISE log file */ 
+        log=fopen("NISE.log","a");
+        fprintf(log,"Finished sample %d\n",samples);
+          
+        time_now=log_time(time_now,log);
+        fclose(log);
+    }/* Closing the loop over samples */
+    
+    /* Divide with total number of samples */
+    for (i=0;i<non->singles;i++){
+        for (j=0;j<non->singles;j++){
+            J[non->singles*i+j]=J[non->singles*i+j]/samples;
+        }
+    }
+
+    /* The calculation is finished, lets write output */
+    log=fopen("NISE.log","a");
+    fprintf(log,"Finished Calculating Response!\n");
+    fprintf(log,"Writing to file!\n");  
+    fclose(log);
+
+    if (non->cluster!=-1){
+        printf("Of %d samples %d belonged to cluster %d.\n",samples,Ncl,non->cluster);
+        if (samples==0){ /* Avoid dividing by zero */ 
+            samples=1;
+        }
+    }
+
+    fclose(mu_traj),fclose(H_traj);
+    if (non->cluster!=-1){
```

Comment:
It seems Cfile is always opened, but not always closed

Reply:
I think there is actually an identical if statement in the open_files function.